packages/scoutgame/src/users/mergeUserAccount.ts

```diff
@@ -0,0 +1,215 @@
+import type { Prisma } from '@charmverse/core/prisma-client';
+import { prisma } from '@charmverse/core/prisma-client';
+
+export type ProfileToKeep = 'current' | 'new';
+
+export const mergeUserAccount = async ({
+  userId,
+  farcasterId,
+  telegramId,
+  selectedProfile
+}: {
+  userId: string;
+  farcasterId?: number;
+  telegramId?: number;
+  selectedProfile: ProfileToKeep;
+}) => {
+  if (!farcasterId && !telegramId) {
+    throw new Error('No account identities to merge');
+  }
+
+  const [primaryUser, secondaryUser] = await Promise.all([
+    prisma.scout.findUniqueOrThrow({
+      where: {
+        id: userId
+      },
+      select: {
+        builderStatus: true
+      }
+    }),
+    prisma.scout.findFirstOrThrow({
+      where: farcasterId ? { farcasterId } : { telegramId },
+      select: {
+        builderStatus: true,
+        id: true
+      }
+    })
+  ]);
+
+  if (primaryUser.builderStatus !== null && secondaryUser.builderStatus !== null) {
+    throw new Error('Can not merge two builder accounts');
+  }
+
+  // The id of the user to retain
+  const retainedUserId = secondaryUser.builderStatus !== null ? secondaryUser.id : userId;
+
+  // The id of the user to merge into the retained user
+  // The merged account must not be a builder
+  const mergedUserId = secondaryUser.builderStatus === null ? secondaryUser.id : userId;
```

Comment:
please put this logic in the frontend, so that this function is more deterministic. Just take a primaryId and secondaryId (or farcaster/telegram of the secondary account). The caller/client should know which userId is being kept, and this function shouldn't just change that fact based on the db state. it would be better to just throw an error if the input is invalid

Reply:
the frontend also needs to know if the user has to be logged out or not (because the button will say "Merge and Log out") so there shouldnt be a reason to put the logic this far down

Reply:
the client does know which user will be kept based on the builder status. the reason primaryId or secondaryId is not passed to the action is because it can be tampered with easily. we already have the primary user which is the logged in user and the secondary user will be obtained from the auth data provided from tg or farcaster.